### DISCUSSION TITLE
SuperSocketLite를 사용하고 게임 개발을 하고 있다면 댓글로 알려주시기 바랍니다^^

### DISCUSSION THREAD

**Discussion by jacking75** [jacking75/SuperSocketLite#19] [General]
SuperSocketLite를 사용하고 게임 개발을 하고 있다면 댓글로 알려주시기 바랍니다^^

**Comment by Z5SteveKim** [comment]
안녕하세요.
실제로 상용화된 게임이 있는지 궁금합니다...
문제점이 없는지도 궁금하네요.

**Comment by jacking75** [comment]
문제점은 없었습니다(제가 아는 한)

이전 회사에서 사용한 경험이 있습니다.
몇몇 분들이 사용했다는 이야기는 들었는데 정확한 정보는 이후 듣지 못해서 정확하게는 모릅니다.
컴투스의 `포커타워디펜스`, `낚시의 신`, `신작 게임`에서 사용하고 있습니다

**Reply by Z5SteveKim** [reply]
답변 감사합니다^^
혹시 분산시스템으로 해서작업한다면 MmoRpg / MoRpg 류의 게임에서도 괜찮을지요?
아직 슈퍼소켓에 대해서 잘 몰라서 질문이 많았습니다 ^^;;
아 모바일 게임으로 작업중입니다.

**Reply by jacking75** [reply]
어떤 게임이냐에 따라 다르겠지만 아주 고성능을 원하는게 아니면 충분히 가능하다고 봅니다.
SuperSocket을 사용하지는 않았지만 출시된  게임 중 MMORPG를 C#으로 만든 경우가 여러 있습니다.

**Reply by Z5SteveKim** [reply]
답변 감사합니다 ^^
디아블로 같은 morpg 게임 서버를 제작해보려는데..
SuperSocket 을 이용해봐야겠네요.^^

**Reply by Z5SteveKim** [reply]
한가지만 더 질문 드리고 싶습니다.

SuperSocket 과 SuperSocketList 는 기본 SuperSocket에서 사용하기 쉽게 감싼 형태로 구현이 되어있는건지요?
차이점이 궁금합니다.

**Reply by jacking75** [reply]
현재 SuperSocket는 2.0 이 개발 중인 것으로 알고 있습니다(아마 거의 완료된). 
SuperSocketLite는 기존 1.6 버전을 개량한 것입니다.
SuperSocket 1.6은 닷넷프레임워크만 지원하고 있는데 SuperSocketLite는 
- 닷넷 코어 지원
- (게임 개발에 필요하지 않는)필요 없는 기능 제거

이 2개가 크게 다릅니다.
특히 게임 개발에서는 필요하지 않을 것으로 예상되는 기능을 제거함으로써 SuperSocket 1.6에 비해 코드가 좀 더 간결해졌습니다.
그렇지만 코어한 부분은 SuperSocket 1.6 과 SuperSocketLite는 동일합니다.

장래에는 SuperSocketLite에 최신 닷넷 기술을 적용하는 것을 계획은 하고 있습니다.

**Reply by Z5SteveKim** [reply]
너무나 친절한 답변 감사합니다 ^^

**Reply by Z5SteveKim** [reply]
아.. 제가 처음 봤을때 SuperSocket Lite 였는데요.. 혹시 list 로 명칭 이 변경 된걸가요..??
^^;;

**Reply by jacking75** [reply]
어이쿠...제가 오랫동안 오타인 상태로 있었네요 ;;;; 알려주셔서 고맙습니다

**Reply by Z5SteveKim** [reply]
아닙니다 ^^;;

한가지 더 여쭈어 보고싶습니다.
튜토리얼 코드를 돌려보고 있습니다.

그런데 pvp game 코드에서 로그인 말고는 다른 어떤 버튼을 눌러도 예외처리 되면서 작동이 안되는듯합니다. ㅠㅜ
혹시 어떤 식으로 진행 되는지 알수 있을지요?
NetworkReadProcess 의 var data = PacketBuffer.Read(); 구문에서 byte[0] 으로 떨어지면서 패킷 파싱 하는곳에서 오류가 나서요..

**Reply by jacking75** [reply]
오래전에 만든 것이라 코드를 좀 봐야 알겠습니다. 그리고 이 문제는 SuperSocketLite의 문제는 아니고 애플리케이션 레이어의 문제라서 클라이언트와 서버의 애플리케이션 코드를 분석하고 이해하셔야 합니다. 이 샘플은 제가 누군가에게 설명하기 위해서 만든 것으로 꽤 오랫동안 사용하지 않아서 시간을 내서 봐야 할 것 같습니다.  다른 질문으로 등록해주시면 실행해보고 나서 답변을 달도록 하겠습니다

**Reply by jacking75** [reply]
이유를 알았습니다. 제가 클라이언트 코드를 완성하지 않았기 때문입니다;;;
PvPGameServer의 클라이언트는 서버와 MessagePack 포맷으로 패킷을 주고 받아야 하는데 로그인만 적용하고, 나머지는 바이너리포맷으로 패킷을 만들고 있어서 발생한 문제였습니다. 
이번주말 안에 수정해서 커밋하도록 하겠습니다^^;

**Reply by Z5SteveKim** [reply]
감사합니다 ^^
제가 테스트로 해보는 프로젝트에는 약간의 수정으로 작동 되도록 해두긴 했습니다만,
튜토리얼 코드올려주시면 다시한번 점검 해보도록할게요 ^^

**Comment by Z5SteveKim** [comment]
안녕하세요.
질문이 있어서 오랜만에 글을 남깁니다.
더미 테스트를 하다가 예외가 발생하는데요. 그럴때 프로그램 자체가 죽어버립니다..
해당 부분을 코드에서 try catch 로 묶어두고 다시 테스트중입니다만..
컨텐츠 단에서 뭔가를 잘못작업이 된걸가요..?
분석하다 질문 한번 드려 봄니다.

`Unhandled exception. System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.Sockets.SocketAsyncEventArgs'.
   at System.Net.Sockets.SocketAsyncEventArgs.ThrowForNonFreeStatus(Int32 status)
   at System.Net.Sockets.SocketAsyncEventArgs.SetBuffer(Byte[] buffer, Int32 offset, Int32 count)
   at SuperSocket.SocketEngine.AsyncSocketSession.ClearPrevSendState(SocketAsyncEventArgs e) in SuperSocketLite\AsyncSocketSession.cs:line 114
   at SuperSocket.SocketEngine.AsyncSocketSession.OnSendingCompleted(Object sender, SocketAsyncEventArgs e) in SuperSocketLite\AsyncSocketSession.cs:line 103
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Net.Sockets.SocketAsyncEventArgs.<>c.<.cctor>b__176_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
   at System.Threading.ThreadPoolTypedWorkItemQueue`2.System.Threading.IThreadPoolWorkItem.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()`

**Reply by jacking75** [reply]
에러 내용으로 봐서는 소켓에서 사용하는 버퍼쪽의 데이터 객체에 문제가 있는 것 같습니다. 
예를들면 메모리 사용을 줄이기 위해서 메모리풀 같은 것을 자체적으로 사용하다가 이 메모리 데이터 객체가 이미 사용이 다 끝난 상태이던가.. 

IO는 비동기로 동작하기 때문에 IO가 완전하게 끝나기 전까지 관련 데이터는 잘 보관 되어야 합니다

**Reply by Z5SteveKim** [reply]
버퍼쪽 데이터 객체라면 패킷 데이터를 말씀하시는건지요.?
패킷 데이터는 I 이던 O 이던 들어올때 new 로 만들어저서 중간에 변경되는 로직이 없는데..
컨텐츠단에서 I/O 등의 데이터를 중간에 조작하거나 직접 release 등의 행위등으로 인해서 문제가 있을지요?
음.. 그렇다면 좀더 로직을 잘 살펴봐야 할거같긴하지만..
제가 말씀하신 내용을 잘 이해 한건지 모르겠습니다.

**Reply by jacking75** [reply]
제가 코드를 알 수가 없으니 정확하게는 말 못하지만 에러 메시지만을 봤을 때는 IO로 사용할 버퍼를 없애버려서 발생하는 에러 같습니다

**Reply by Z5SteveKim** [reply]
네, 말씀 감사합니다.

**Comment by seeingistrue** [comment]
안녕하세요~~
올려주신 turorial을 일부 수정하여 각 프로젝트의 통신쪽 기본 코드로 사용하려 합니다.
게임쪽은 아니고 여러가지 embedded 장비와 연결하는 pc쪽 테스트 GUI APP에 사용할 예정입니다.
그리고 수정된 코드를 제 개인 github에 올리려하는데 가능할까요?
제가 github에 아직 능숙하지 않아, 온라인상에서 참조하기 위함인데 혹시 
다른 방법이 있으면 알려주실수 있을까요?
감사합니다.~

**Reply by jacking75** [reply]
안녕하세요
오픈소스라서 마음대로 사용 하셔도 괜찮습니다^^

**Reply by seeingistrue** [reply]
공개해주신 소스가 큰 도움이 되고 있습니다.
감사합니다~~

**Comment by yonjuhhee** [comment]
잘 쓰고 있습니다. 감사합니다.

**Reply by jacking75** [reply]
고맙습니다^^
